### PR TITLE
Optimize PMTU probing process

### PR DESCRIPTION
**Pull request by sairoop-elear**

### CODE REVIEW

File: src/net_packet.c [hunk 1]
```diff
@@ -67,23 +67,10 @@ static void send_mtu_probe_handler(event_loop_t *loop, void *data) {
...
-		n->mtuprobes = 1;
-		n->minmtu = 0;
-		n->maxmtu = MTU;
-	}
-
-	if(n->mtuprobes >= 10 && n->mtuprobes < 32 && !n->minmtu) {
+	if(n->mtuprobes >= 10 && n->mtuprobes < 30 && !n->minmtu) {
 		logger(mesh, MESHLINK_INFO, "No response to MTU probes from %s", n->name);
-		n->mtuprobes = 31;
+		n->mtuprobes = 10;
```
Comment (gsliepen) [inline src/net_packet.c]: But this means that if we don't get a reaction to PMTU probes, we keep sending a burst of three random sized packets, but now just every `mesh->pinginterval`. I think in this case we should also send the smallest possible probe; when we get a response we will restart the PMTU discovery from scratch anyway.

File: test/blackbox/run_blackbox_tests/run_blackbox_tests.c [hunk 2]
```diff
@@ -89,7 +89,7 @@ int main(int argc, char *argv[]) {
 	failed_tests += test_meshlink_set_status_cb();
 	failed_tests += test_meshlink_join();
 	failed_tests += test_meshlink_set_channel_poll_cb();
-//  failed_tests += test_meshlink_channel_open_ex();
+	//  failed_tests += test_meshlink_channel_open_ex();
```
Comment (gsliepen) [inline test/blackbox/run_blackbox_tests/run_blackbox_tests.c]: Just a comment: it's better not to add unrelated changes into a pull request.

File: src/net_packet.c [hunk 3]
```diff
@@ -443,6 +441,11 @@ bool receive_sptps_record(void *handle, uint8_t type, const void *data, uint16_t
 
 	vpn_packet_t inpkt;
 
+	if(type == PKT_PING) {
+		logger(mesh, MESHLINK_DEBUG, "Got MTU probe length %d from %s", len, from->name);
+		return true;
```
[the remark this answers is not in the document]
Comment (sairoop-elear) [reply src/net_packet.c]: I modified the code which can do request and reply flow for pinging compatible with the original design but with fewer probes and small probe size.

File: src/graph.c [hunk 4]
```diff
@@ -224,7 +224,7 @@ static void check_reachability(meshlink_handle_t *mesh) {
 
 			n->status.udp_confirmed = false;
 			n->maxmtu = MTU;
-			n->minmtu = 0;
+			n->minmtu = 500;
```
Comment (gsliepen) [inline src/graph.c]: Setting `n->minmtu` to a non-zero value indicates to the rest of the code that UDP communication of packets up to that size is possible, so this is bad. If we want to ensure we don't send MTU probes smaller than 500 bytes, we should just change `send_mtu_probe_handler()` to only send probes bigger than 500 bytes.

File: src/net_packet.c [hunk 5]
```diff
@@ -102,7 +99,14 @@ static void send_mtu_probe_handler(event_loop_t *loop, void *data) {
 		timeout = mesh->pinginterval;
 		goto end;
 	} else if(n->mtuprobes == 32) {
+		vpn_packet_t packet;
+		packet.probe = true;
+		packet.len = 1;
+
+		logger(mesh, MESHLINK_DEBUG, "Sending MTU probe length %d to %s", packet.len, n->name);
+		sptps_send_record(&n->sptps, PKT_PROBE, packet.data, packet.len);
```
Comment (gsliepen) [inline src/net_packet.c]: This looks OK!

File: src/net_packet.c [hunk 6]
```diff
@@ -154,6 +158,23 @@ void send_mtu_probe(meshlink_handle_t *mesh, node_t *n) {
 static void mtu_probe_h(meshlink_handle_t *mesh, node_t *n, vpn_packet_t *packet, uint16_t len) {
 	logger(mesh, MESHLINK_DEBUG, "Got MTU probe length %d from %s", packet->len, n->name);
 
+	if(n->mtuprobes > 30) {
+		n->status.udp_confirmed = true;
```
Comment (gsliepen) [inline src/net_packet.c]: You must not set `status.udp_confirmed` to `true` for probe requests, only for probe replies. The reason is that we want to test for bidirectional communication. I would just not touch anything in `mtu_probe_h()`.

**Comment by sairoop-elear** [comment]
I made those above commented changes.

File: src/net_packet.c [hunk 7]
```diff
@@ -385,6 +382,11 @@ bool send_sptps_data(void *handle, uint8_t type, const void *data, size_t len) {
 			to->incompression = mesh->self->incompression;
 			return send_request(mesh, to->nexthop->connection, "%d %s %s %s -1 -1 -1 %d", ANS_KEY, mesh->self->name, to->name, buf, to->incompression);
 		} else {
+
+			if(to->mtu && (to->mtuprobes > 30)) {
```
Comment (gsliepen) [inline src/net_packet.c]: Why do you reset `to->mtuprobes` here?

File: src/net_packet.c [hunk 8]
```diff
@@ -409,12 +411,22 @@ bool send_sptps_data(void *handle, uint8_t type, const void *data, size_t len) {
...
+			if(to->mtu && (to->mtuprobes > 30) && !type) {
+				to->mtuprobes = 1;
+			}
 		} else {
 			logger(mesh, MESHLINK_WARNING, "Error sending UDP SPTPS packet to %s: %s", to->name, sockstrerror(sockerrno));
 			return false;
 		}
 	}
 
+	if(to->mtu && (to->mtuprobes > 30) && !type) {
```
Comment (gsliepen) [inline src/net_packet.c]: This doesn't feel right. We should not need to reset the ping timer because we are trying to send packets. This would also mean that when we are sending data frequently (more than once every ping interval), we keep resetting the timer before it ever expires, and then the MTU handlers will not run.

Comment (sairoop-elear) [reply src/net_packet.c]: One of our optimizations that we planned in PMTU probing process is that If there are some data packets being sent to the other node at the end then those packets are also keeping NAT mappings alive and so we can skip the next ping probe(s) and done by resetting MTU time handler with `mesh->pinginterval` value.

File: src/net_packet.c (hunk 7 above)
Comment (sairoop-elear) [reply src/net_packet.c]: In one of our discussions we deduced that 2 instances or triggers where we can find sending via UDP failed are 
1. When we redirect UDP data via TCP
2. When ICMP sends message too big response for the UDP packet sent
Hence there we can rediscover the PMTU process and verify PMTU size.

**Comment by gsliepen** [comment]
> In one of our discussions we deduced that 2 instances or triggers where we can find sending via UDP failed are 
> 1. When we redirect UDP data via TCP
> 2. When ICMP sends message too big response for the UDP packet sent
> Hence there we can rediscover the PMTU process and verify PMTU size.

1. is wrong, it's the other way around: if we know we can't send via UDP
we will redirect data via TCP.

-- 
Met vriendelijke groet / with kind regards,
     Guus Sliepen <guus@tinc-vpn.org>

**Comment by sairoop-elear** [comment]
outdated
